src/kronecker.jl

```diff
@@ -0,0 +1,258 @@
+is a `LinearMap` object among the first 8 arguments.
+
+# Examples
+```jldoctest; setup=(using LinearAlgebra, SparseArrays, LinearMaps)
+julia> J = LinearMap(I, 2) # 2×2 identity map
+LinearMaps.UniformScalingMap{Bool}(true, 2)
+
+julia> E = spdiagm(-1 => trues(1)); D = LinearMap(E + E' - 2I);
+
+julia> Δ₁ = kron(D, J) + kron(J, D); # discrete 2D-Laplace operator
+
+julia> Δ₂ = LinearMaps.kronsum(D, D); # same operator as Kronecker sum
+
+julia> Matrix(Δ₁) == Matrix(Δ₂)
+true
+```
+"""
+kronsum(A::LinearMap{TA}, B::LinearMap{TB}) where {TA,TB} = KroneckerSumMap{promote_type(TA,TB)}((A, B))
+kronsum(A::LinearMap, B::LinearMap, C::LinearMap, Ds::LinearMap...) = kronsum(A, kronsum(B, C, Ds...))
+# promote AbstractMatrix arguments to LinearMaps, then take LinearMap-Kronecker sum
+for k in 1:8 # is 8 sufficient?
+    Is = ntuple(n->:($(Symbol(:A,n))::AbstractMatrix), Val(k-1))
+    # yields (:A1, :A2, :A3, ..., :A(k-1))
+    L = :($(Symbol(:A,k))::LinearMap)
+    # yields :Ak
+    mapargs = ntuple(n -> :(LinearMap($(Symbol(:A,n)))), Val(k-1))
+    # yields (:LinearMap(A1), :LinearMap(A2), ..., :LinearMap(A(k-1)))
+
+    @eval kronsum($(Is...), $L, As::Union{LinearMap,AbstractMatrix}...) =
+        kronsum($(mapargs...), $(Symbol(:A,k)), promote_to_lmaps(As...)...)
+end
+
+Base.size(A::KroneckerSumMap, i) = prod(size.(A.maps, i))
+Base.size(A::KroneckerSumMap) = (size(A, 1), size(A, 2))
+
+LinearAlgebra.issymmetric(A::KroneckerSumMap) = all(issymmetric, A.maps)
+LinearAlgebra.ishermitian(A::KroneckerSumMap{<:Real}) = all(issymmetric, A.maps)
+LinearAlgebra.ishermitian(A::KroneckerSumMap) = all(ishermitian, A.maps)
+
+LinearAlgebra.adjoint(A::KroneckerSumMap{T}) where {T} = KroneckerSumMap{T}(map(adjoint, A.maps))
+LinearAlgebra.transpose(A::KroneckerSumMap{T}) where {T} = KroneckerSumMap{T}(map(transpose, A.maps))
+
+Base.:(==)(A::KroneckerSumMap, B::KroneckerSumMap) = (eltype(A) == eltype(B) && A.maps == B.maps)
+
+function LinearMaps.A_mul_B!(y::AbstractVector, L::KroneckerSumMap, x::AbstractVector)
+    A, B = L.maps
+    ma, na = size(A)
+    mb, nb = size(B)
+    (length(y) == size(L, 1) && length(x) == size(L, 2)) || throw(DimensionMismatch("kronecker product"))
+    X = reshape(x, (nb, na))
+    Y = reshape(y, (nb, na))
+    mul!(Y, X, convert(AbstractMatrix, transpose(A)))
+    mul!(Y, B, X, true, true)
+    return y
+end
+
+LinearMaps.At_mul_B!(y::AbstractVector, A::KroneckerSumMap, x::AbstractVector) = A_mul_B!(y, transpose(A), x)
+
+LinearMaps.Ac_mul_B!(y::AbstractVector, A::KroneckerSumMap, x::AbstractVector) = A_mul_B!(y, adjoint(A), x)
+
+"""
+    kron(A::LinearMap, k::Int)
+
+Construct a lazy representation of the `k`-th Kronecker power `A ⊗ A ⊗ ... ⊗ A`
+for `k≥2`. This function is currently not type-stable!
+"""
+function Base.kron(A::LinearMap, k::Int)
+    k > 1 || throw(ArgumentError("the Kronecker power is only defined for exponents larger than 1, got $k"))
+    return kron(Base.fill_to_length((), A, Val(k))...)
+end
```

Comment:
I would assume that type stability is ok when `k` is a constant in the source code, i.e. a literal integer.

However, where is this function/definition coming from. Also its position in the source code is very strange, since this is about Kronecker product, not kronecker sum, but it also seems like something that should not be a method of `kron`. This seems analoguous to defining `*(a::Number, k::Int) = a^k`?

Reply:
This is a convenience functionality that I "copied" from Kronecker.jl. There, however, additional types (for the Kronecker power and the Kronecker sum power) are introduced, for which then special mixed-product/multiplication rules are introduced etc. Since everything is completely lazy here (and there, actually), I thought there is no reason to introduce new types, but to "blow up" `A` to `(A, A, ..., A)` and to pass it to `kron`. We could change the function name and make it, say, `kronpow` or so, but this adds to the number of exported functions, which I wanted to keep small. But I really don't have strongly feelings either way.

In a sense, it is analogous to what you're saying, except that we don't have an alternative notation for the right hand side in the Kronecker context.

Reply:
I still think that this is in violation of what functions and different methods thereof are supposed to represent, according to the Julia guidelines/common practice. A function should have one specific functionality, albeit this being implemented in different ways for different types. These are two distinct types of functionality, kronecker product and kronecker power. In particular, because there is a perfectly valid interpretation to the kronecker product of a matrix and a number. In fact, in Base (actually, standard library LinearAlgebra)
`kron( ::Matrix, ::Number)` is implemented, and does what it means mathematically, i.e. `M ⊗ s = M * s`, with `M` matrix and `s` a scalar.  

So I would indeed suggest a new method. Another possibility is some unicode hackery and exporting `⊗`, though technically that is also a new function.

```julia
julia> struct KronWrap{T<:Integer}
               p::T
       end

julia> ⊗(k::Integer) = KronWrap(k)
⊗ (generic function with 1 method)

julia> ⊗(a, b, c...) = kron(a, b, c...)
⊗ (generic function with 2 methods)

julia> Base.:^(a::Matrix, p::KronWrap) = kron(ntuple(n->a, p.p)...)

julia> a = randn(2,2)
2×2 Array{Float64,2}:
 -1.16944    0.141176
  0.590445  -1.12108 

julia> a^⊗(3)
8×8 Array{Float64,2}:
 -1.59933    0.193071    0.193071   -0.0233076   0.193071   -0.0233076  -0.0233076   0.00281371
  0.807491  -1.53319    -0.0974806   0.185087   -0.0974806   0.185087    0.0117679  -0.0223438 
  0.807491  -0.0974806  -1.53319     0.185087   -0.0974806   0.0117679   0.185087   -0.0223438 
 -0.407697   0.774098    0.774098   -1.46979     0.0492173  -0.0934494  -0.0934494   0.177433  
  0.807491  -0.0974806  -0.0974806   0.0117679  -1.53319     0.185087    0.185087   -0.0223438 
 -0.407697   0.774098    0.0492173  -0.0934494   0.774098   -1.46979    -0.0934494   0.177433  
 -0.407697   0.0492173   0.774098   -0.0934494   0.774098   -0.0934494  -1.46979     0.177433  
  0.205844  -0.390837   -0.390837    0.742086   -0.390837    0.742086    0.742086   -1.40901   

julia> a ⊗ a ⊗ a
8×8 Array{Float64,2}:
 -1.59933    0.193071    0.193071   -0.0233076   0.193071   -0.0233076  -0.0233076   0.00281371
  0.807491  -1.53319    -0.0974806   0.185087   -0.0974806   0.185087    0.0117679  -0.0223438 
  0.807491  -0.0974806  -1.53319     0.185087   -0.0974806   0.0117679   0.185087   -0.0223438 
 -0.407697   0.774098    0.774098   -1.46979     0.0492173  -0.0934494  -0.0934494   0.177433  
  0.807491  -0.0974806  -0.0974806   0.0117679  -1.53319     0.185087    0.185087   -0.0223438 
 -0.407697   0.774098    0.0492173  -0.0934494   0.774098   -1.46979    -0.0934494   0.177433  
 -0.407697   0.0492173   0.774098   -0.0934494   0.774098   -0.0934494  -1.46979     0.177433  
  0.205844  -0.390837   -0.390837    0.742086   -0.390837    0.742086    0.742086   -1.40901   
```

Reply:
Yeah, that's cool. I would call that struct `KronPower`, to indicate that `⊗(3)` is to be understood as a Kronecker power, in contrast to a regular multiplicative power. Moreover, we can use the same trick with the Kronecker sum power, via exporting `⊕` and introducing `KronSumPower`.

Reply:
Indeed. I have to admit that I had never heard of a kronecker sum before (even though I use/encounter this structure often), but find it very confusing that it is the same symbol as a direct sum, which is clearly very different.

Reply:
Hm, I just did the change. Do you think there will be any conflict with the direct sum if we start exporting `\oplus`? It seems pretty common in the "Kronecker community" to use it for the Kronecker sum.

Reply:
I did not even know there was a "Kronecker community" :-). Good documentation is key, I guess. There is no dedicated direct sum in Base either, and it just seems to fit in the wider context of block matrices, so I think it's fine to have `\oplus` be Kronecker sum.

Reply:
> I did not even know there was a "Kronecker community" :-).

Maybe I should've used "matrix product community". 😉 There are many other similar matrix products people seem to use in, for instance, signal processing and other fields, see https://en.wikipedia.org/wiki/Kronecker_product#Related_matrix_operations

One thing to think about is how we want to (at some point) construct block diagonal maps. For that purpose, \oplus might be handy as well.

What I like a lot about `LinearMaps.jl` pre-Kronecker product is that we only export `LinearMap`, and everything else is overloaded from Base/LinearAlgebra, in contrast to what is done in other similar packages. That allows for nice generic programming, without the need to use dedicated functions which work for LinearMaps exclusively. But I guess we may want to overload `cat(As...; dims=(1,2))` for block diagonal concatenation, so using oplus for Kronecker sum (with clear documentation) should be indeed fine.

Reply:
`SparseArrays.jl` uses `blockdiag` which seems OK, though `Diagonal(As...)` would be a logical extension of `Diagonal(v)` and seems more natural/readable than `cat(...)`.
Right now my PR #65 uses `block_diag` because I was getting error messages about conflicts with `SparseArrays.blockdiag` that I could not resolve, but that's off topic.
I've never seen `\oplus` used for block diagonals.
I have no concerns about a new export because one could do `using LinearMaps: LinearMap` to get just the desired functions.